contracts/stake/src/state.rs

```diff
@@ -70,204 +67,214 @@ impl StakeState {
+            .get_stake_mut(&account)
             .expect("A stake should exist in the map to be unstaked!");
+        let prev_stake = Some(*loaded_stake);
 
-        let prev_value = Some(loaded_stake.clone());
+        // ensure there is a value staked, and that the withdrawal is exactly
+        // the same amount
+        let staked_value = loaded_stake
+            .amount
+            .expect("There must be an amount to unstake")
+            .value;
 
-        let counter = loaded_stake.counter();
-
-        let (value, _) = loaded_stake.remove_amount();
-        loaded_stake.increment_counter();
-
-        // verify signature
-        let digest =
-            Unstake::signature_message(counter, value, unstake.address)
-                .to_vec();
+        if value != staked_value {
+            panic!("Value withdrawn different from staked amount");
+        }
 
-        if !rusk_abi::verify_bls(digest, unstake.public_key, unstake.signature)
-        {
+        // check signature is correct
+        let digest = unstake.signature_message().to_vec();
+        if !rusk_abi::verify_bls(digest, account, signature) {
             panic!("Invalid signature!");
         }
-        // make call to transfer contract to withdraw a note from this contract
-        // containing the value of the stake
-        let withdraw_data = Mint {
-            value,
-            address: unstake.address,
-            sender: unstake.public_key,
-        };
+
+        // make call to the transfer contract to withdraw funds from this
+        // contract into the receiver specified by the withdrawal.
         let _: () =
-            rusk_abi::call(TRANSFER_CONTRACT, "withdraw", &withdraw_data)
-                .expect(
-                "Withdrawing the balance from contract should be successful",
-            );
+            rusk_abi::call(TRANSFER_CONTRACT, "withdraw", transfer_withdraw)
+                .expect("Withdrawing stake should succeed");
+
+        // update the state accordingly
+        loaded_stake.amount = None;
 
         rusk_abi::emit(
             "unstake",
-            StakingEvent {
-                public_key: unstake.public_key,
-                value,
+            StakeEvent {
+                account,
+                value: staked_value,
+                receiver: Some(*transfer_withdraw.receiver()),
             },
         );
 
-        let key = unstake.public_key.to_bytes();
+        let key = account.to_bytes();
         self.previous_block_state
             .entry(key)
-            .or_insert((prev_value, unstake.public_key));
+            .or_insert((prev_stake, account));
     }
 
     pub fn withdraw(&mut self, withdraw: Withdraw) {
-        // deplete the stake from a key and increment the signature counter
-        let loaded_stake = self
-            .get_stake_mut(&withdraw.public_key)
-            .expect("A stake should exist in the map to be withdrawn!");
+        self.check_new_block();
```

Comment:
This is not needed since withdraw is not changing the provisioners staked amount.
If you really want to add it to be more consistent, please update `previous_block_state` accordingly at the end of the method

Reply:
This is a holdover from removing the stake changes. Will remove.